File: app/src/test/java/com/dougkeen/bart/data/FollowedTripStoreTest.java
package com.dougkeen.bart.data;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import com.dougkeen.bart.model.Departure;
import com.dougkeen.bart.model.Line;
import com.dougkeen.bart.model.Station;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.nio.file.Files;

public class FollowedTripStoreTest {
    private File directory;
    private File storageFile;
    private FollowedTripStore store;

    @Before
    public void setUp() throws Exception {
        directory = Files.createTempDirectory("bart-followed-trip").toFile();
        storageFile = new File(directory, "followed_trip.json");
        store = new FollowedTripStore(storageFile);
    }

    @After
    public void tearDown() {
        delete(directory);
    }

    @Test
    public void saveAndLoadRestoresTheTripAcrossStoreInstances() {
        Departure original = departure();

        store.save(original);
        Departure restored = new FollowedTripStore(storageFile).load();

        assertEquals(Station.MONT, restored.getOrigin());
        assertEquals(Station.RICH, restored.getPassengerDestination());
        assertEquals(Line.RED, restored.getLine());
        assertEquals("trip-1", restored.getTripLegs().get(0).getTripId());
        assertTrue(storageFile.exists());
    }

    @Test
    public void clearingRemovesTheDurableState() {
        store.save(departure());
        store.save(null);

        assertFalse(storageFile.exists());
        assertNull(store.load());
    }

    @Test
    public void malformedStateIsDiscardedInsteadOfRetriedForever() throws Exception {
        Files.write(storageFile.toPath(), "not-json".getBytes());

        assertNull(store.load());
        assertFalse(storageFile.exists());
    }

    private static Departure departure() {
        Departure departure = new Departure();
        departure.setOrigin(Station.MONT);
        departure.setTrainDestination(Station.RICH);
        departure.setPassengerDestination(Station.RICH);
        departure.setLine(Line.RED);
        departure.setDirection("n");
        departure.setPlatform("2");
        departure.setMinEstimate(4_000_000_000L);
        departure.setMaxEstimate(4_000_060_000L);

        com.dougkeen.bart.model.TripLeg leg = new com.dougkeen.bart.model.TripLeg(
                Line.RED, Station.MONT, Station.RICH, Station.RICH, "trip-1",
                0L, 0L, java.util.Collections.emptyList());
        departure.setTripLegs(java.util.Collections.singletonList(leg));
        return departure;
    }

    private static void delete(File file) {
        if (file == null || !file.exists()) {
            return;
        }
        File[] children = file.listFiles();
        if (children != null) {
            for (File child : children) {
                delete(child);
            }
        }
        file.delete();
    }
}

File: app/src/test/kotlin/in/izyum/bart/data/FollowedTripStoreTest.kt
package `in`.izyum.bart.data

import `in`.izyum.bart.model.Departure
import `in`.izyum.bart.model.Itinerary
import `in`.izyum.bart.model.Line
import `in`.izyum.bart.model.Station
import `in`.izyum.bart.model.TripLeg
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import java.io.File
import java.nio.file.Files

class FollowedTripStoreTest {
    private lateinit var directory: File
    private lateinit var storageFile: File
    private lateinit var store: FollowedTripStore

    @Before
    fun setUp() {
        directory = Files.createTempDirectory("bart-followed-trip").toFile()
        storageFile = File(directory, "followed_trip.json")
        store = FollowedTripStore(storageFile)
    }

    @After
    fun tearDown() {
        delete(directory)
    }

    @Test
    fun saveAndLoadRestoresAnItineraryAcrossStoreInstances() {
        val itinerary = Itinerary.fromDeparture(departure())!!
        store.saveItinerary(itinerary)

        val restored = FollowedTripStore(storageFile).loadItinerary()!!

        assertEquals(itinerary.origin, restored.origin)
        assertEquals(itinerary.destination, restored.destination)
        assertEquals("trip-1", restored.legs.single().tripId)
    }

    @Test
    fun clearingRemovesTheDurableState() {
        store.saveItinerary(Itinerary.fromDeparture(departure()))
        store.saveItinerary(null)
        assertFalse(storageFile.exists())
        assertNull(store.loadItinerary())
    }

    @Test
    fun malformedStateIsDiscardedInsteadOfRetriedForever() {
        Files.write(storageFile.toPath(), "not-json".toByteArray())
        assertNull(store.loadItinerary())
        assertFalse(storageFile.exists())
    }

    private fun departure(): Departure = Departure.builder()
        .setOrigin(Station.MONT).setTrainDestination(Station.RICH)
        .setPassengerDestination(Station.RICH).setLine(Line.RED)
        .setPlatform("2")
        .setMinEstimate(4_000_000_000L).setMaxEstimate(4_000_060_000L)
        .setTripLegs(listOf(TripLeg(
            Line.RED, Station.MONT, Station.RICH, Station.RICH,
            "trip-1", 0L, 0L, emptyList(),
        ))).build()

    private fun delete(file: File?) {
        if (file == null || !file.exists()) return
        file.listFiles()?.forEach(::delete)
        file.delete()
    }
}
